runtime/webhookserver/webhooktunnel.go

```diff
@@ -0,0 +1,100 @@
+package webhookserver
+
+import (
+	"errors"
+	"net/http"
+	"sync"
+
+	"github.com/taskcluster/slugid-go/slugid"
+	"github.com/taskcluster/taskcluster-client-go"
+	"github.com/taskcluster/taskcluster-client-go/auth"
+	"github.com/taskcluster/webhooktunnel/util"
+	"github.com/taskcluster/webhooktunnel/whclient"
+)
+
+// WebhookTunnel wraps a whclient instance as a WebHookServer
+type WebhookTunnel struct {
+	m        sync.RWMutex
+	handlers map[string]http.Handler
+	// url is assigned when the client connects to the proxy
+	client *whclient.Client
+}
+
+// NewWebhookTunnel returns a pointer to a new WebhookTunnel instance
+func NewWebhookTunnel(credentials *tcclient.Credentials) (*WebhookTunnel, error) {
+	configurer := func() (whclient.Config, error) {
+		authClient := auth.New(credentials)
+		whresp, err := authClient.WebhooktunnelToken()
+		if err != nil {
+			return whclient.Config{}, errors.New("could not get token from tc-auth")
+		}
+
+		// This hack is needed since proxyURL in auth config is set as a http url
+		proxyURL := whresp.ProxyURL
+		proxyURL = util.MakeWsURL(proxyURL)
+
+		return whclient.Config{
+			ID:        whresp.TunnelID,
+			ProxyAddr: proxyURL,
+			Token:     whresp.Token,
+			UseDomain: true,
+		}, nil
+	}
+
+	client, err := whclient.New(configurer)
+	if err == whclient.ErrAuthFailed {
+		client, err = whclient.New(configurer)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	wt := &WebhookTunnel{
+		handlers: make(map[string]http.Handler),
+		client:   client,
+	}
+
+	go func() {
+		http.Serve(wt.client, http.HandlerFunc(wt.handle))
+	}()
+	return wt, nil
+}
+
+// AttachHook adds a new webhook to the server
+func (wt *WebhookTunnel) AttachHook(handler http.Handler) (string, func()) {
+	id := slugid.Nice()
+	wt.m.Lock()
+	wt.handlers[id] = handler
+	wt.m.Unlock()
+
+	url := wt.client.URL() + "/" + id + "/"
+	detach := func() {
+		wt.m.Lock()
+		defer wt.m.Unlock()
+		delete(wt.handlers, id)
+	}
+
+	return url, detach
+}
+
+// Stop will close the webhooktunnel client
+func (wt *WebhookTunnel) Stop() {
+	_ = wt.client.Close()
```

Comment:
I always like to be explicit to the reader that "No, I didn't forget to check the return error value, I just don't care about it" +1